pkg/live/inventoryrg.go

```diff
@@ -249,78 +308,432 @@ func (icm *InventoryResourceGroup) GetObject() (*unstructured.Unstructured, erro
+			return err
+		}
+		if !exists {
+			continue
+		}
+
+		newItems := []interface{}{}
+		for _, item := range items {
+			fields, ok := item.(map[string]interface{})
+			if !ok {
+				return fmt.Errorf("unable to cast nested fields to the required types for: %v", item)
+			}
+
+			id, err := idFromFieldInterface(fields)
+			if err != nil {
+				return err
+			}
+
+			if _, exists := managedResources[id]; exists {
+				newItems = append(newItems, item)
+			} else {
+				needCollapse = true
+			}
+		}
+		err = unstructured.SetNestedSlice(rg.UnstructuredContent(), newItems, "spec", "resources")
+		if err != nil {
+			return err
+		}
+		_, err = namespacedClient.Update(context.TODO(), &rg, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+	}
+
+	if needCollapse {
+		// Reshuffle GKNNs to the start of ResourceGroup object series.
+		return icm.collapse(dc, mapper)
+	}
+
+	return nil
 }
 
-func (icm *InventoryResourceGroup) ApplyWithPrune(dc dynamic.Interface, mapper meta.RESTMapper, statusPolicy inventory.StatusPolicy, _ object.ObjMetadataSet) error {
-	invInfo, namespacedClient, err := icm.getNamespacedClient(dc, mapper)
+// collapse is a utility function to rebalance objects stored within sharded ResourceGroups.
+func (icm *InventoryResourceGroup) collapse(dc dynamic.Interface, mapper meta.RESTMapper) error {
+	if ResourceCount == DisableSharding {
+		klog.V(4).Info("skipping re-balancing of ResourceGroups as sharding is disabled")
+		return nil
+	}
+
+	namespacedClient, err := icm.getNamespacedClient(dc, mapper)
 	if err != nil {
 		return err
 	}
 
-	// Update the cluster inventory object.
-	appliedObj, err := namespacedClient.Update(context.TODO(), invInfo, metav1.UpdateOptions{})
+	// Get all cluster ResourceGroup objects using label selector.
+	resourceList, err := namespacedClient.List(context.TODO(), metav1.ListOptions{
+		LabelSelector: fmt.Sprintf(shardedRGLabelFmt, icm.Name()),
+	})
+	if err != nil {
+		return err
+	}
+
+	resourcegroups := resourceList.Items
+	sortResourceGroups(resourcegroups)
+
+	// Get all GKNNs across all sharded ResourceGroups.
+	items := []interface{}{}
+	for _, rg := range resourcegroups {
+		resources, exists, err := unstructured.NestedSlice(rg.Object, "spec", "resources")
+		if err != nil || !exists {
+			continue
+		}
 
-	// Update status.
-	if statusPolicy == inventory.StatusPolicyAll {
-		invInfo.SetResourceVersion(appliedObj.GetResourceVersion())
-		_, err = namespacedClient.UpdateStatus(context.TODO(), invInfo, metav1.UpdateOptions{})
+		items = append(items, resources...)
 	}
 
-	return err
+	count := len(items)
+
+	// Re-balance GKNN objects in sharded ResourceGroups.
+	objIdx := 0
+	for objIdx < count {
+		var curr []interface{}
+		if objIdx+ResourceCount < len(items) {
+			curr = items[objIdx : objIdx+ResourceCount]
+		} else {
+			curr = items[objIdx:]
+		}
+		err := unstructured.SetNestedSlice(resourcegroups[objIdx/ResourceCount].UnstructuredContent(), curr, "spec", "resources")
```

Comment:
This seems unsafe. Maybe pull out `resourcegroups[objIdx/ResourceCount]` and/or `objIdx/ResourceCount` to variables?